### PR TITLE
TL review of day 4 homework

### PR DESCRIPTION
**Pull request by tomlatham**
The code mostly works as expected, except for one missed edge case - adding Z to the end of text that already ends with Z.  However, the applyCipher code is not very readable and is roughly a factor 2 - 3 slower than the implementation that uses maps when compiling without optimisation.  When activating optimisation level 2, this rises to nearly a factor 10.
When reviewing that part of the code I have not changed it to use maps but have simply tried to make it more readable (although these changes also result in very slightly faster performance due to the reduction in the number of branches).
I suggest that, if you have time, you try to do the implementation in terms of maps and compare the performance yourself.  I have done so by encrypting the dictionary file on my computer, which takes around 2.5 - 3 seconds with your code depending on the level of compiler optimisation and between 0.3 and 1.1 seconds for ours.
You had not included any test of the PlayfairCipher, so I have added a rather simple example.  To really check all edge cases would require a more detailed set of tests.
Please see the line-by-line comments for more details.  Let me know if you have questions.

### CODE REVIEW

File: MPAGSCipher/PlayfairCipher.cpp [hunk 1]
```diff
@@ -1,7 +1,5 @@
 // Standard library includes
-#include <iostream>
 #include <string>
-#include <vector>
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Remove unnecessary includes.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 2]
```diff
@@ -35,8 +33,8 @@ void PlayfairCipher::setKey( const std::string& key )
     // Change J -> I
     auto func2 = [] (char in_char)
     {
-        if(in_char=='J') return('I');
-        else return(in_char);
+        if(in_char=='J') {return('I');}
+        else {return(in_char);}
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Always best to include braces even on one-line conditionals to avoid unexpected issues cropping up in future - see the Apple SSL "goto fail" bug.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 3]
```diff
@@ -87,50 +88,64 @@ std::string PlayfairCipher::applyCipher( const std::string& inputText, const Cip
 
     if(outputText.size()%2 != 0)
     {
-        outputText.push_back('Z');
+        outputText += (outputText.back() == 'Z') ? 'X' : 'Z';
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: If the text already ends with a Z you need to add some other letter, in this case we've picked X.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 4]
```diff
@@ -87,50 +88,64 @@ std::string PlayfairCipher::applyCipher( const std::string& inputText, const Cip
...
-        int index1 = key_.find(outputText[i]);
-        int index2 = key_.find(outputText[i+1]);
+        size_t index1 = key_.find(outputText[i]);
+        size_t index2 = key_.find(outputText[i+1]);
+
+        size_t row1 = index1/5;
+        size_t row2 = index2/5;
+
+        size_t column1 = index1%5;
+        size_t column2 = index2%5;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: For clarity, decompose the global index into row and column indices.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 5]
```diff
@@ -87,50 +88,64 @@ std::string PlayfairCipher::applyCipher( const std::string& inputText, const Cip
...
 
         //   - Apply the rules to these coords to get 'new' coords
-        if( (index2/5-index1/5)==0 )
+        if( row1 == row2 )
         {
+            // Row - so increment/decrement the column indices (modulo 5)
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: The rule conditions now become rather clearer.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 6]
```diff
@@ -87,50 +88,64 @@ std::string PlayfairCipher::applyCipher( const std::string& inputText, const Cip
...
             if(cipherMode==CipherMode::Encrypt)
             {
-                outputText[i]   = key_[index1+((index1%5==4)?(-4):(1))];
-                outputText[i+1] = key_[index2+((index2%5==4)?(-4):(1))];
+                column1 = (column1 + 1) % 5;
+                column2 = (column2 + 1) % 5;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Similarly, the actions become clearer (and do not involve branches).

File: MPAGSCipher/PlayfairCipher.cpp [hunk 7]
```diff
@@ -87,50 +88,64 @@ std::string PlayfairCipher::applyCipher( const std::string& inputText, const Cip
...
         {
-            outputText[i]   = key_[index1+(index2%5-index1%5)];
-            outputText[i+1] = key_[index2-(index2%5-index1%5)];
+            // Rectangle/Square - so keep the rows the same and swap the columns
+            std::swap( column1, column2 );
         }
 
         //   - Find the letter associated with the new coords
+        index1 = 5*row1 + column1;
+        index2 = 5*row2 + column2;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Finally, recalculate the global index from the row and column indices.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 8]
```diff
@@ -87,50 +88,64 @@ std::string PlayfairCipher::applyCipher( const std::string& inputText, const Cip
...
+
+        //   - Make the replacements
+        outputText[i]   = key_[index1];
+        outputText[i+1] = key_[index2];
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Then get the corresponding letter from the key.
